Hawkular/add ssl support

josejulio opened the pull request:
Adds support for tls endpoints on Hawkular MW provider

ToDo:
 - [x] Rename `container_security_options` to `endpoint_security_options`

![screenshot from 2017-02-23 11-15-26](https://cloud.githubusercontent.com/assets/3845764/23273321/464563ec-f9c4-11e6-8799-6bac3b0392fe.png)

![screenshot from 2017-02-23 11-26-29](https://cloud.githubusercontent.com/assets/3845764/23273325/4bd04d04-f9c4-11e6-95f0-262340609ba9.png)

![screenshot from 2017-02-23 11-16-02](https://cloud.githubusercontent.com/assets/3845764/23273331/52f39c80-f9c4-11e6-9c6e-21ea13e087f4.png)



This PR ~~depends on PR #450 and~~ supports https://github.com/ManageIQ/manageiq/pull/14054

@miq-bot add-label providers/hawkular

josejulio commented:
@cben I modified some of your work, could you please take a look and let me know if I should duplicate the code or is OK what i did?

josejulio commented on app/controllers/mixins/ems_common_angular.rb:
```diff
@@ -432,15 +458,29 @@ def set_ems_record_vars(ems, mode = nil)
+    def container_security_options(security_protocol, certificate_authority)
+      {
+        :security_protocol     => security_protocol,
+        :verify_ssl            => %w(ssl-without-validation non-ssl).exclude?(security_protocol),
```
@cben 
Added `non-ssl` to the options that shouldn't be verified by ssl.

cben replied:
Nice.  Not all providers will use same logic, but good to share.
Let's rename the function to something more generic, say `endpoint_security_options`?

josejulio replied:
Sure. Will do once your PR is merged. (adding it as a todo).

josejulio commented on app/views/layouts/angular-bootstrap/_endpoints_angular.html.haml:
```diff
@@ -114,6 +116,59 @@
+%div{"ng-if" => defined?(tls_ca_certs_hide) ? false : true}
+  .form-group{"ng-if"=> "emsCommonModel.emstype == 'rhevm' || " +                                          |
+                        "((emsCommonModel.ems_controller == 'ems_container' || " +                         |
+                        "emsCommonModel.emstype == 'hawkular') && " +                               |
```
Also using this `Trusted CA Certificates` textarea.

josejulio commented:
@miq-bot add-label WIP

cben commented on app/views/layouts/angular-bootstrap/_endpoints_angular.html.haml:
```diff
@@ -143,7 +153,8 @@
   .form-group{"ng-if"=> "emsCommonModel.emstype == 'rhevm' || " +                                          |
-                        "(emsCommonModel.ems_controller == 'ems_container' && " +                          |
+                        "((emsCommonModel.ems_controller == 'ems_container' || " +                         |
+                        "emsCommonModel.emstype == 'hawkular') && " +                               |
```
This is perhaps too nested, hard to read.
One option that might be more readable if you put the custom condition first (untested):
```
                         "(emsCommonModel.#{prefix}_security_protocol == 'ssl-with-validation-custom-ca' &&" |
                         " (emsCommonModel.ems_controller == 'ems_container' || " +                          |
                         "  emsCommonModel.emstype == 'hawkular'))"}                                         |

However the meaning of security_protocol is provider-dependent and this file already does a lot of copy-pasting to keep different providers independent.
What do you think of doing separate (container && custom) || (hawkular && custom)?

josejulio replied:
Grouping them different seems a good option that would give more flexibility to modify them as needed for each provider without having to copy/paste the whole block.

josejulio commented:
@miq-bot add_label middleware, wip, pending core

josejulio commented:
@miq-bot rm_label wip

josejulio commented:
#450 is merged already.
@cben I made the changes you suggested. As i modified some of your code, could you please review again? Thanks!

abonas commented:
Wouldn't it be better to first put the protocol field, and then after the user selects it, to fill automatically the default port? (which user can of course change)

josejulio commented:
@abonas I'm reusing UI from https://github.com/ManageIQ/manageiq-ui-classic/pull/450 

What you propose is possible but would end being inconsistent with other providers as they define the port before the security protocol.

https://github.com/josejulio/manageiq-ui-classic/blob/b8d8736225787281808e52b8c7a9688d8f07e1d8/app/views/layouts/angular-bootstrap/_endpoints_angular.html.haml#L27-L59

I could move our `security_protocol` field and fill the `port` field if that's not a concern..

cben commented:
I think some providers do reset port based on security protocol:
https://github.com/ManageIQ/manageiq-ui-classic/blob/d3045891dff33b70db46a27ac8dbe067d1aaf55c/app/assets/javascripts/controllers/ems_common/ems_common_form_controller.js#L360-L368
in which case yes, having protocol above port makes sense.

Don't assume #450 is perfect.  I missed several spots there, and still chasing a bug there.  Will cc you as soon I have a fix.

josejulio commented:
If that makes sense for both of you, I'll move the port below the security protocol and fill default values.

josejulio commented:
I added the code to fill the port depending on selected security protocol.

I'm not sure on the order of the elements. We have the following (in all providers):

1. Host
2. Port
3. Database_name
4. Security_Protocol
5. Verify TLS Certificates
6. Trusted CA Certificates

I'm not sure if put it like this (a):

1. Host
2. Security_Protocol
3. Port
4. Database_name
5. Verify TLS Certificates
6. Trusted CA Certificates

or this (b):

1. Host
2. Security_Protocol
3. Verify TLS Certificates
4. Trusted CA Certificates
5. Port
6. Database_name


I suppose (a) would makes more sense to have Host and port not so far apart. 
WDYT?

cben commented:
Out of these options I also prefer (a), host and port shouldn't be distant.

Note _endpoints_angular.html.haml is shared, reordering this will affect all providers.  Perhaps belongs in separate PR.

But yes having 3rd control reset what one types in 2nd control can be annoying.  (In fact, it may be annoying whatever the order, people don't always fill top-down.)
What if changing the protocol only rewrote the port if it's == the other standard port?  I mean change 8443 -> 8080 and 8080 -> 8443 but leave other values alone.

@mansam you did the Openstack logic for protocol automatically setting port. Any opinion on field order, and/or softer logic that leaves uncommon port values alone?

josejulio commented:
@miq-bot rm_label pending core

josejulio commented:
I agree about having it in a separate PR, others might want to take a look and review that change.

I changed the logic so it will only fill if the port is currently empty (first time), has a validation error (undefined) or is any of the default ports.

mansam commented:
I agree with @cben that given the two options, a) is the better one since it keeps host and port together and hopefully reduces the frustration of the form wiping out their port entry.

I would suggest only allowing the form to automatically fill the port on protocol change if the field is empty (because the user cleared it to get the default back), or if the field value is identical to the default, indicating that the user hasn't changed it. Any other value should be treated as something deliberate the user entered, so it should be kept. (The Openstack form unfortunately isn't very smart about this.) If you do this, you can probably afford to keep Security Protocol after Port, but putting it first is fine either way.

josejulio commented:
Need travis to run tests again. It failed installing dependencies.

josejulio commented:
Thank you @mansam. Did some changes to avoid overriding user entered values.

cben commented:
See just opened #670, most fixes probably also apply to you.

AparnaKarve commented:
Like @cben said above, changes in #670 are possibly applicable here too.

josejulio commented:
@AparnaKarve I guess I could apply these once #670 is merged (In a separate PR)

https://github.com/ManageIQ/manageiq-ui-classic/pull/670/files#diff-f402e411cd6dc158c724b14d72ead612R324

AparnaKarve commented:
>  I could apply these once #670 is merged (In a separate PR if possible)

@josejulio, sure, that works!

abonas commented:
@serenamarie125 and @Loicavenel , please have a look at the UX issue in regards of fields order for protocol, port, etc. Current way of things don't make a lot of sense (and causes a bit more work for the user). Please decide how it should look in all the providers.

Loicavenel commented:
We should at minimum reverse Port and Protocol where it does exist. Some providers does not have Port but use URL as input. If protocols has different port, then default port should be presented and updated baed on the protocol selection

josejulio commented:
@abonas @Loicavenel 
Should i do that on this PR? 
Please see https://github.com/ManageIQ/manageiq-ui-classic/pull/460#issuecomment-285138289
There might be other fields between Port and Security Protocol.



> If protocols has different port, then default port should be presented and updated baed on the protocol selection

Currently (on this provider) the Port is only updated (when selecting a protocol) IF port is empty OR port is any of the default ports OR port is invalid (e.g. NaN).

cben commented:
Since order is shared to all providers, I think separate PR is natural.

P.S. another order you haven't listed is put protocol even before hostname.
 (Kinda like proto://host:port URL structure.)

josejulio commented:
That would be the third option! I'll edit https://github.com/ManageIQ/manageiq-ui-classic/pull/460#issuecomment-285138289

josejulio commented:
@AparnaKarve Updated code so Trusted CA Certificates now affects required validation.

Next steps would be:
Pass down a default value of `default_security_protocol` to the angular controller using [this](https://github.com/ManageIQ/manageiq-ui-classic/pull/670/files#diff-f402e411cd6dc158c724b14d72ead612R324) after is merged and preferably on a separate PR.
Change port / security protocol order (IMO (and @cben [opinion](https://github.com/ManageIQ/manageiq-ui-classic/pull/460#issuecomment-286479273)) belongs in a separate PR)

Is there anything else you want me to address on this PR?

AparnaKarve commented:
> Updated code so Trusted CA Certificates now affects required validation.

@josejulio Good call on that.

Agree about addressing the other issues you mentioned in separate PRs.
And with that, this is GTG.

josejulio commented:
@AparnaKarve wait a bit. I think i missed something.

josejulio commented:
@AparnaKarve Ok, done, I had an extra ´}´ on the last commit. Just removed it.

josejulio commented:
Ummh, @AparnaKarve https://github.com/ManageIQ/manageiq-ui-classic/pull/670 also address the Trusted CA certificates, so I'm removing it from this PR because is the same code.

abonas commented:
@josejulio @cben please see @Loicavenel guideline regarding the protocol/port order, it could be applied in this PR or in a separate one, although I don't see much point doing it here in the incorrect way and then redoing it again in another PR.

AparnaKarve commented:
@josejulio #670 has some code that you would need to fix this completely, so how about we wait for #670 to get merged first? (which I'm currently reviewing)

Once it's merged, it will be easy for you to identify what else needs to be fixed from your end.

josejulio commented:
@AparnaKarve Sure: I can wait for it.

@abonas We are sharing the security_protocol with [others](https://github.com/josejulio/manageiq-ui-classic/blob/7413410b1db9f7c109cf20565f08c2a1068b7e94/app/views/layouts/angular-bootstrap/_endpoints_angular.html.haml#L83-L89) providers. I just want to make sure I don't break anything from their end. By doing a separate PR i could at least try to have them look at only that fix. 
It shouldn't break anything though.

josejulio commented:
@AparnaKarve 

Rebased and added the `security_protocol_default` for `ems_middleware`
Now when editing a previously added provider it shows `Non SSL` selected
![screenshot from 2017-03-16 18-49-13](https://cloud.githubusercontent.com/assets/3845764/24024486/d71292b4-0a81-11e7-8b1f-66f60bcb4092.png)

AparnaKarve commented:
@josejulio The screenshot looks good. So looks like you're all set now.

cben commented:
LGTM :+1:

abonas commented:
@miq-bot assign @martinpovolny please review & merge

abonas commented:
@miq-bot assign @martinpovolny
@martinpovolny please review and merge
